src/Voron/Data/Containers/Container.cs

```diff
@@ -316,29 +312,280 @@ public string Dump()
+                if (Unsafe.Add(ref currentMetadata, index).IsFree == false)
+                    return true;
+            
             return false;
         }
 
-        public int SpaceUsed()
+        /// <summary>
+        /// Calculates space used in all container items.
+        /// Equals operation forEachMetadata.Sum(i => container.MetadataFor(i).Get(ptr)).
+        /// </summary>
+        /// <param name="pagePtr">Storage of size</param>
+        /// <param name="usedItems">Number of items that are currently in use.</param>
+        /// <returns></returns>
+        internal int SpaceUsedInItems(byte* pagePtr, out int usedItems)
         {
             int numberOfOffsets = Header.NumberOfOffsets;
-            var size = numberOfOffsets * sizeof(ItemMetadata) + PageHeader.SizeOf;
-            for (int i = 0; i < numberOfOffsets; i++)
+            var size = 0;
+            usedItems = numberOfOffsets;
+            var index = 0;
+
+            Debug.Assert(Vector512.IsHardwareAccelerated == false || Vector512<ushort>.Count == 32);
+            //Vector512<ushort>.Count; explicit declaration when hardware does not support Vector
+            ushort* offsetArray = stackalloc ushort[32]; 
+            
+            if (AdvInstructionSet.IsAcceleratedVector512)
             {
-                size += MetadataFor(i).GetSize(_page.Pointer);
+                var N = Vector512<ushort>.Count;
+                for (; index + N <= numberOfOffsets; index += N)
+                {
+                    var metadataItems = Vector512.Load((ushort*)_page.DataPointer + index);
+                    var offsets = Vector512.BitwiseAnd(
+                        left: (metadataItems >> ItemMetadata.OffsetShift), 
+                        right: Vector512.Create(ItemMetadata.OffsetMask));
+                    offsets.Store(offsetArray);
+                    
+                    var sizes = Vector512.BitwiseAnd(
+                        left: metadataItems, 
+                        right: Vector512.Create(ItemMetadata.SizeMask));
+                    
+                    var byteElements = Vector512.Equals(
+                        left: sizes, 
+                        right: Vector512.Create(ItemMetadata.ByteSizeElement));
+                    
+                    var ushortElements = Vector512.Equals(
+                        left: sizes, 
+                        right: Vector512.Create(ItemMetadata.UshortSizeElement));
+                    
+                    var byteSizedPopulation = byteElements.ExtractMostSignificantBits();
+                    var ushortSizedPopulation = ushortElements.ExtractMostSignificantBits();
+                    var usedPopulation = byteSizedPopulation | ushortSizedPopulation;
+                    var zeroPopulation = Vector512.Equals(sizes, Vector512<ushort>.Zero);
+                    usedItems -= BitOperations.PopCount(zeroPopulation.ExtractMostSignificantBits());
+                    
```

Comment:
Let's add some documentation on how this algorithm works. All the bit trickery is great until you have to work with it 10 years down the line. Don't ask me how I know. 🤦‍♂️